2022/Day7/Solve.py

```diff
@@ -0,0 +1,83 @@
+class Day_Six(object):
+  def day_six(self):
+    computer = self.parse("PuzzleInput.txt")
+    computer.calculate_size()
+    print("The answer to part one is", computer.sum_all_subdirectories_under_limit(100000))
+
+  def parse(self, puzzle_input):
+    current_directory = []
+    computer = Directory("dir /")
+    with open(puzzle_input, "r") as f:
+      for line in f:
+        if "cd /" in line:
+          continue
+        if ".." in line:
+          current_directory.pop()
+          continue
+        elif "$ cd" in line:
+          current_directory.append("dir " + line.split()[2])
+          continue
+        elif "$" not in line:
+          if "dir" in line:
+            item = Directory(line.strip())
+          else:
+            item = File(line)
+          computer.add_files(current_directory, item)
+
+    return computer
+
+
+class Directory(object):
+  name: str
+  size: int
+  contents: {}
+
+  def __init__(self, name):
+    self.name = name
+    self.contents = {}
+    self.size = 0
+
+  def calculate_size(self):
+    for key in self.contents.keys():
+      if "dir" in key:
```

Comment:
You're not leveraging your interface very well here - why doesn't `get_size` on a directory just calculate its true size?

Reply:
Oof, also - what happens if someone calls calculate_size() multiple times?  this function would return something different every time...

Reply:
So this is a challenge I ran into, that I didn't know how to solve. `get_size` did originally do that, but then when I called `sum_all_subdirectories_under_limit` it did exactly what you said in your second comment. It kept making everything bigger and bigger. This was my hack around. I called it once, got the sizes calculated, and then could just return size.

How do I better do this so it doesn't change size if called `get_size` multiple times?

Reply:
What if a file name happens to be generated with the letter combination of "dir" in it?

Reply:
fair point

Reply:
Why store a size at all? Why not define get_size like:

```python
  def get_size(self):
    size = 0
    for f in self.contents.values():
      size = size + f.get_size()
    return size
```

Reply:
Because I thought that this would slow things down, honestly, and so chose to store the size to speed things up?

Reply:
Yeah - let AoC prove to you that performance updates are necessary.

Also if you wanted to memoize, you could instead:
```python
def get_size(self):
  if self.size == -1:
    self.size = sum([f.get_size() for f in self.contents.values()])
  return self.size
  
def add_file(self, path, new_file):
  self.size == -1
  ... # actually add the file
  ```